### DISCUSSION TITLE
I need to send data to the server. I want to ask some questions

### DISCUSSION THREAD

**Discussion by greenking19** [simolus3/drift#1901] [Q&A]
I now use drift to store local data, which makes me a lot more portable!
Now I need to pass the data to the server.
My app usage logic is like this 👇
1. data exists locally
2. Script detects unsynced data and uploads
3. After the upload is successful, mark this Row as uploaded. Failure is not marked

I use "updateStatus" to mark whether to upload, I have a dozen tables like this.
The task data is often modified, and every time I modify it, I will mark` updateStatus = "0"`

0. not upload
1. update success

```dart
class TaskTable extends Table{
  ........
  // 0. not upload
  // 1. update success
  TextColumn get updateStatus => text()();
}
```
Because the original intention of my app is not to be affected by the network, let the operation continue. So I choose to store the data locally, and then use the script to upload the data.
My current method is in the loop, one by one table to check the data that has not been uploaded. Then upload one by one
Below is my sync code, it feels so stupid.
I summarize a few key points
1. data is placed locally
2. timely upload to the backend
3. uploads may not be successful, so check often
```dart
Future<void> syncData() async {
  List<TaskTableData> taskList = await (db.select(db.task)
        ..where((tbl) => tbl.updateStatus.equals('0')))
      .get();
  if (taskList.length > 0) {
    await uploadTask(taskList.first);
  }
  await Future.delayed(Duration(milliseconds: 100));
  syncData();
}

Future<void> uploadTask(TaskTableData task) {
  postToServer(tasl);
}
```
I have more than a dozen tables, I feel too bloated to write it all like this.
But I can't think of any other way 🥲
At the beginning I wanted to use `db.select(xxx).watch()`, But it doesn't seem to match my needs.
This has bothered me for a long time 🥲
So I want to ask for your opinion to see if there is a good solution  ☺️

**Comment by simolus3** [comment]
So unfortunately I'm not aware of an easy way to query all updatable tables in a single query. However, you can reduce the amount of duplicate code a bit. First, you can extract the column's definition into a mixin:

```dart
mixin WithUpdateStatus on Table {
  TextColumn get updateStatus => text()(); // by the way, this could also be a BoolColumn ;)
}

class TaskTable extends Table with WithUpdateStatus {
  // other columns
}

// other tables also mixing in WithUpdateStatus
```

And then you could do something like:

```dart
Future<void> syncData() async {
  for (final table in allTables) {
    if (table is WithUpdateStatus) {
      final entries = await (db.select(table)..where((_) => table.updateStatus.equals('0'))).get();
      await postToServer(entries);  // this method would have to check which entries it got (so do a chain of if (x is TaskTableData) etc
    }
  }
  await Future.delayed(Duration(milliseconds: 100));
  syncData();
}
```

IIRC there's a virtual stacktrace between async functions, so you probably don't want to have that infinite recursion there. Alternatively, you could write it with an infinite loop:

```dart
// just call this method and don't await it
Future<Never> startSynchronizationLoop() async {
  while (true) {
    await Future.delayed(Duration(milliseconds: 100));
    await syncData();
  }
}
```

You can also re-run the synchronization logic whenever tables change (but in this case I really recommend something like [debounceTime](https://pub.dev/documentation/rxdart/latest/rx/DebounceExtensions/debounceTime.html) to avoid this running too foten):

```dart
Future<Never> startLoop() async {
  // Note: You should use something like debounce from rxdart to avoid this
  // stream updating too often.
  final updates = tableUpdates(TableUpdateQuery.allOf([
    for (final table in allTables)
      if (table is WithSchemaUpdate) TableUpdateQuery.onTable(table),
  ])).debounceTime(const Duration(seconds: 2));

  await for (final update in updates) {
    await syncData();
  }
}
```

In my opinion this could form the base for a reasonably simple synchronization logic if you only need to go from the app to the server (and there can be no synchronization conflicts). Some things to consider:

- You probably want some retry logic if uploading something fails.
- Also, you should pause the synchronization loop while the app is in the background.

**Reply by greenking19** [reply]
Thank you!
I Learned to lot. 
I hope I can hole others in the future!
I want to buy you a cup of coffee☕️ , but I can't seem to find the donation address. 
Can you provide an address? 😊

**Reply by greenking19** [reply]
Andotid Studio doesn't know about `table with WithSchemaUpdate`😕

![image](https://user-images.githubusercontent.com/43332372/175807461-e33dfbb2-eb3f-4324-962e-c6ed6bbf9967.png)
I'm now trying other function implementations

**Reply by greenking19** [reply]
I tried to manually write all the tables `with WithSchemaUpdate` into a Map.
Then I use `table.actualTableName` to exchange the desired table,
but it doesn't work when I use `where`

**Reply by simolus3** [reply]
> Andotid Studio doesn't know about `table with WithSchemaUpdate`

Ah right. I think `(table as WithSchemaUpdate).updateStatus.equals('0')` might work?

**Reply by greenking19** [reply]
Thanks, this solved my problem perfectly. 
Thanks a lot for your help 🤓

The native data is synced to the server, and the headache for me is that I have a web project that manages the data. I now have to judge whether the data has been deleted elsewhere. 
Now I am implementing it.
